File: kernel/skills/loader.py
"""SKILL.md parser — loads Agent Skills spec files into structured manifests.

Format (per https://agentskills.io/specification):

    ---
    name: skill-name
    description: What it does and when to use it.
    license: MIT
    compatibility: Requires Python 3.12+
    metadata:
      author: someone
      version: "1.0"
    allowed-tools: Read Write Bash(git:*)
    ---

    # Skill body (Markdown)

    Instructions for the agent to follow when this skill is active.

Directory structure:

    skill-name/
    ├── SKILL.md          # Required
    ├── scripts/          # Optional: executable code
    ├── references/       # Optional: detailed docs
    └── assets/           # Optional: static resources
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml

from kernel.skills.validator import ValidationResult, validate_frontmatter

logger = logging.getLogger(__name__)

_FRONTMATTER_DELIM = "---"


@dataclass
class SkillManifest:
    """A parsed SKILL.md file.

    Attributes:
        name: Unique skill identifier (matches directory name).
        description: What the skill does and when to use it.
        body: Markdown instructions for the agent.
        license: Optional license name or reference.
        compatibility: Optional environment requirements.
        metadata: Optional arbitrary key-value pairs.
        allowed_tools: Optional space-separated tool allowlist.
        skill_dir: Absolute path to the skill directory.
        scripts_dir: Path to scripts/ if exists.
        references_dir: Path to references/ if exists.
        assets_dir: Path to assets/ if exists.
        source: Origin identifier — "builtin", "user", "catalog:anthropic", etc.
    """

    name: str
    description: str
    body: str
    license: str | None = None
    compatibility: str | None = None
    metadata: dict[str, Any] = field(default_factory=dict)
    allowed_tools: str | None = None
    skill_dir: Path = field(default_factory=lambda: Path("."))
    scripts_dir: Path | None = None
    references_dir: Path | None = None
    assets_dir: Path | None = None
    source: str = "unknown"

    @property
    def has_scripts(self) -> bool:
        return self.scripts_dir is not None and self.scripts_dir.exists()

    @property
    def has_references(self) -> bool:
        return self.references_dir is not None and self.references_dir.exists()

    @property
    def has_assets(self) -> bool:
        return self.assets_dir is not None and self.assets_dir.exists()

    @property
    def tool_list(self) -> list[str]:
        """Parse allowed-tools into a list. Empty if None."""
        if not self.allowed_tools:
            return []
        return self.allowed_tools.split()

    def to_dict(self) -> dict[str, Any]:
        """Serialize to JSON-safe dict (for API responses)."""
        return {
            "name": self.name,
            "description": self.description,
            "body": self.body,
            "license": self.license,
            "compatibility": self.compatibility,
            "metadata": dict(self.metadata),
            "allowed_tools": self.allowed_tools,
            "allowed_tools_list": self.tool_list,
            "skill_dir": str(self.skill_dir),
            "has_scripts": self.has_scripts,
            "has_references": self.has_references,
            "has_assets": self.has_assets,
            "source": self.source,
        }


class SkillParseError(ValueError):
    """Raised when SKILL.md cannot be parsed."""


def _split_frontmatter(content: str) -> tuple[dict[str, Any], str]:
    """Extract YAML frontmatter and Markdown body from raw file content.

    Returns:
        Tuple of (frontmatter_dict, body_markdown).

    Raises:
        SkillParseError: If frontmatter delimiters are missing or YAML is invalid.
    """
    lines = content.splitlines(keepends=True)
    if not lines:
        raise SkillParseError("SKILL.md is empty")

    # First non-empty line must be the opening delimiter
    first_idx = 0
    while first_idx < len(lines) and not lines[first_idx].strip():
        first_idx += 1

    if first_idx >= len(lines) or lines[first_idx].strip() != _FRONTMATTER_DELIM:
        raise SkillParseError(
            "SKILL.md must begin with YAML frontmatter delimited by '---'"
        )

    # Find closing delimiter
    end_idx = None
    for i in range(first_idx + 1, len(lines)):
        if lines[i].strip() == _FRONTMATTER_DELIM:
            end_idx = i
            break

    if end_idx is None:
        raise SkillParseError(
            "Unterminated frontmatter: missing closing '---'"
        )

    frontmatter_text = "".join(lines[first_idx + 1 : end_idx])
    body = "".join(lines[end_idx + 1 :]).lstrip("\n")

    try:
        data = yaml.safe_load(frontmatter_text) or {}
    except yaml.YAMLError as exc:
        raise SkillParseError(f"Invalid YAML frontmatter: {exc}") from exc

    if not isinstance(data, dict):
        raise SkillParseError(
            f"Frontmatter must be a mapping, got {type(data).__name__}"
        )

    return data, body


def load_skill(
    skill_dir: Path | str,
    *,
    source: str = "unknown",
    strict: bool = True,
) -> SkillManifest:
    """Load a SKILL.md file from a skill directory.

    Args:
        skill_dir: Path to skill directory (must contain SKILL.md).
        source: Origin tag (e.g. "builtin", "user", "catalog:anthropic").
        strict: If True, raise ValidationError on spec violations.
                If False, log warnings but return best-effort manifest.

    Returns:
        Parsed SkillManifest.

    Raises:
        FileNotFoundError: If SKILL.md is missing.
        SkillParseError: If file is malformed.
        ValidationError: If strict=True and validation fails.
    """
    skill_dir = Path(skill_dir).resolve()
    skill_md = skill_dir / "SKILL.md"

    if not skill_md.is_file():
        raise FileNotFoundError(f"SKILL.md not found in {skill_dir}")

    content = skill_md.read_text(encoding="utf-8")
    frontmatter, body = _split_frontmatter(content)

    # Validate against spec
    result: ValidationResult = validate_frontmatter(
        frontmatter, expected_name=skill_dir.name
    )

    if result.warnings:
        for w in result.warnings:
            logger.warning("Skill '%s': %s", skill_dir.name, w)

    if not result.valid:
        msg = f"Invalid SKILL.md in {skill_dir}: {'; '.join(result.errors)}"
        if strict:
            from kernel.skills.validator import ValidationError
            raise ValidationError("frontmatter", msg)
        logger.warning(msg)

    # Resolve optional subdirectories
    scripts_dir = skill_dir / "scripts"
    refs_dir = skill_dir / "references"
    assets_dir = skill_dir / "assets"

    manifest = SkillManifest(
        name=str(frontmatter.get("name", skill_dir.name)),
        description=str(frontmatter.get("description", "")),
        body=body,
        license=frontmatter.get("license"),
        compatibility=frontmatter.get("compatibility"),
        metadata=dict(frontmatter.get("metadata") or {}),
        allowed_tools=frontmatter.get("allowed-tools"),
        skill_dir=skill_dir,
        scripts_dir=scripts_dir if scripts_dir.is_dir() else None,
        references_dir=refs_dir if refs_dir.is_dir() else None,
        assets_dir=assets_dir if assets_dir.is_dir() else None,
        source=source,
    )

    return manifest


def save_skill(manifest: SkillManifest, skill_dir: Path | str | None = None) -> Path:
    """Write a SkillManifest back to SKILL.md format.

    Used by the converter and publishing flow.

    Args:
        manifest: The manifest to serialize.
        skill_dir: Override target directory (defaults to manifest.skill_dir).

    Returns:
        Path to the written SKILL.md file.
    """
    target_dir = Path(skill_dir) if skill_dir else manifest.skill_dir
    target_dir.mkdir(parents=True, exist_ok=True)

    frontmatter: dict[str, Any] = {
        "name": manifest.name,
        "description": manifest.description,
    }
    if manifest.license:
        frontmatter["license"] = manifest.license
    if manifest.compatibility:
        frontmatter["compatibility"] = manifest.compatibility
    if manifest.metadata:
        frontmatter["metadata"] = dict(manifest.metadata)
    if manifest.allowed_tools:
        frontmatter["allowed-tools"] = manifest.allowed_tools

    yaml_text = yaml.safe_dump(
        frontmatter,
        sort_keys=False,
        allow_unicode=True,
        default_flow_style=False,
    ).rstrip()

    content = f"---\n{yaml_text}\n---\n\n{manifest.body.lstrip()}"
    skill_md = target_dir / "SKILL.md"
    skill_md.write_text(content, encoding="utf-8")
    return skill_md

File: src-tauri/src/backend/skills/loader.rs
//! SKILL.md parser — reads YAML frontmatter + Markdown body and
//! returns a structured [`SkillManifest`].
//!
//! Format (per <https://agentskills.io/specification>):
//!
//! ```text
//! ---
//! name: skill-name
//! description: What it does and when to use it.
//! license: MIT                   # optional
//! compatibility: ...             # optional
//! metadata:                       # optional, free-form mapping
//!   author: someone
//!   version: "1.0"
//! allowed-tools: Read Write Bash(git:*)   # optional, space-separated
//! ---
//!
//! # Skill body
//! ...
//! ```
//!
//! The Rust port mirrors `kernel/skills/loader.py::load_skill`. JSON
//! shape returned by `to_json` matches `SkillManifest.to_dict` so the
//! UI sees identical responses regardless of which backend serves
//! `/skills/installed`.

use std::path::{Path, PathBuf};

use anyhow::{anyhow, Context, Result};
use serde::Deserialize;
use serde_json::{json, Value};

const FRONTMATTER_DELIM: &str = "---";

/// A parsed SKILL.md file. Field set mirrors Python's `SkillManifest`
/// dataclass — keep the JSON serialisation in `to_json` aligned with
/// `kernel/skills/loader.py::SkillManifest.to_dict`.
#[derive(Debug, Clone)]
pub struct SkillManifest {
    pub name: String,
    pub description: String,
    pub body: String,
    pub license: Option<String>,
    pub compatibility: Option<String>,
    pub metadata: Value,
    pub allowed_tools: Option<String>,
    pub skill_dir: PathBuf,
    pub has_scripts: bool,
    pub has_references: bool,
    pub has_assets: bool,
    pub source: String,
}

impl SkillManifest {
    /// Space-separated `allowed-tools` parsed into a list. Empty when
    /// the field is absent or only whitespace.
    pub fn tool_list(&self) -> Vec<String> {
        self.allowed_tools
            .as_deref()
            .map(|s| s.split_whitespace().map(String::from).collect())
            .unwrap_or_default()
    }

    /// JSON shape returned by the `/skills/installed` route. Mirrors
    /// `SkillManifest.to_dict` from the Python module byte-for-byte
    /// (modulo path string formatting on Windows).
    pub fn to_json(&self) -> Value {
        json!({
            "name": self.name,
            "description": self.description,
            "body": self.body,
            "license": self.license,
            "compatibility": self.compatibility,
            "metadata": self.metadata,
            "allowed_tools": self.allowed_tools,
            "allowed_tools_list": self.tool_list(),
            "skill_dir": self.skill_dir.to_string_lossy(),
            "has_scripts": self.has_scripts,
            "has_references": self.has_references,
            "has_assets": self.has_assets,
            "source": self.source,
        })
    }
}

/// Load and parse a SKILL.md from `skill_dir`. The file MUST exist;
/// callers are expected to have probed for it (`registry.rs` does).
/// `source` is propagated into the manifest as a provenance tag
/// (`"user"`, `"builtin"`, `"catalog:<id>"`, etc).
pub fn load_skill(skill_dir: &Path, source: &str) -> Result<SkillManifest> {
    let skill_md = skill_dir.join("SKILL.md");
    let content = std::fs::read_to_string(&skill_md)
        .with_context(|| format!("read SKILL.md at {:?}", skill_md))?;
    let (frontmatter, body) = split_frontmatter(&content)?;

    let dir_name = skill_dir
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("")
        .to_string();
    let name = frontmatter.name.unwrap_or(dir_name);
    let description = frontmatter.description.unwrap_or_default();
    let metadata = frontmatter
        .metadata
        .map(yaml_to_json)
        .transpose()?
        .unwrap_or_else(|| json!({}));

    Ok(SkillManifest {
        name,
        description,
        body,
        license: frontmatter.license,
        compatibility: frontmatter.compatibility,
        metadata,
        allowed_tools: frontmatter.allowed_tools,
        skill_dir: skill_dir.to_path_buf(),
        has_scripts: skill_dir.join("scripts").is_dir(),
        has_references: skill_dir.join("references").is_dir(),
        has_assets: skill_dir.join("assets").is_dir(),
        source: source.to_string(),
    })
}

/// Strict frontmatter shape used internally for the YAML deserialise
/// step. Optional everywhere — required-field validation is the
/// validator's job (Chunk 4); the loader is forgiving so a missing
/// `name` falls back to the directory name (matches Python).
#[derive(Debug, Default, Deserialize)]
pub(crate) struct Frontmatter {
    pub(crate) name: Option<String>,
    pub(crate) description: Option<String>,
    pub(crate) license: Option<String>,
    pub(crate) compatibility: Option<String>,
    pub(crate) metadata: Option<serde_yaml::Value>,
    #[serde(rename = "allowed-tools")]
    pub(crate) allowed_tools: Option<String>,
}

/// Convert the YAML `metadata` mapping into a JSON value so callers
/// can serialise it through serde_json without a double-walk. Errors
/// out only on inputs YAML accepts but JSON cannot represent (e.g.
/// non-string mapping keys). In practice metadata is always
/// `{string: scalar}` and this is a no-op cost-wise.
fn yaml_to_json(value: serde_yaml::Value) -> Result<Value> {
    serde_json::to_value(&value).context("convert YAML metadata to JSON")
}

/// Split file contents into (frontmatter, body). Mirrors the Python
/// `_split_frontmatter` semantics:
///   - Skip leading blank lines.
///   - First non-blank line MUST be `---`.
///   - Find the next `---` line; everything between is YAML.
///   - Body is everything after the closing delim, with leading
///     newlines trimmed.
pub(crate) fn split_frontmatter(content: &str) -> Result<(Frontmatter, String)> {
    let lines: Vec<&str> = content.lines().collect();
    if lines.is_empty() {
        return Err(anyhow!("SKILL.md is empty"));
    }

    let first_idx = lines
        .iter()
        .position(|l| !l.trim().is_empty())
        .ok_or_else(|| anyhow!("SKILL.md is empty"))?;

    if lines[first_idx].trim() != FRONTMATTER_DELIM {
        return Err(anyhow!(
            "SKILL.md must begin with YAML frontmatter delimited by '---'"
        ));
    }

    let end_idx = lines
        .iter()
        .enumerate()
        .skip(first_idx + 1)
        .find(|(_, l)| l.trim() == FRONTMATTER_DELIM)
        .map(|(i, _)| i)
        .ok_or_else(|| anyhow!("Unterminated frontmatter: missing closing '---'"))?;

    let frontmatter_text = lines[first_idx + 1..end_idx].join("\n");
    let body_text = lines[end_idx + 1..].join("\n");
    let body = body_text.trim_start_matches('\n').to_string();

    let parsed: Frontmatter = if frontmatter_text.trim().is_empty() {
        // Empty frontmatter is allowed in lenient mode — matches Python's
        // `data = yaml.safe_load(...) or {}`.
        Frontmatter::default()
    } else {
        serde_yaml::from_str(&frontmatter_text).context("Invalid YAML frontmatter")?
    };

    Ok((parsed, body))
}

/// Test helper: a frontmatter-only Manifest builder. Used by registry
/// integration tests to assemble a `SkillManifest` without touching the
/// filesystem.
#[cfg(test)]
#[allow(dead_code)]
pub(crate) fn manifest_for_test(name: &str, description: &str, source: &str) -> SkillManifest {
    SkillManifest {
        name: name.into(),
        description: description.into(),
        body: String::new(),
        license: None,
        compatibility: None,
        metadata: json!({}),
        allowed_tools: None,
        skill_dir: PathBuf::from("."),
        has_scripts: false,
        has_references: false,
        has_assets: false,
        source: source.into(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn split_frontmatter_minimal() {
        let raw = "---\nname: x\ndescription: y\n---\n\nBody.\n";
        let (fm, body) = split_frontmatter(raw).unwrap();
        assert_eq!(fm.name.as_deref(), Some("x"));
        assert_eq!(fm.description.as_deref(), Some("y"));
        assert_eq!(body, "Body.");
    }

    #[test]
    fn split_frontmatter_skips_leading_blank_lines() {
        let raw = "\n\n---\nname: a\n---\nBody";
        let (fm, body) = split_frontmatter(raw).unwrap();
        assert_eq!(fm.name.as_deref(), Some("a"));
        assert_eq!(body, "Body");
    }

    #[test]
    fn split_frontmatter_rejects_missing_open_delim() {
        let raw = "name: x\ndescription: y\n";
        let err = split_frontmatter(raw).unwrap_err();
        assert!(err.to_string().contains("must begin with YAML frontmatter"));
    }

    #[test]
    fn split_frontmatter_rejects_unterminated() {
        let raw = "---\nname: x\ndescription: y\n";
        let err = split_frontmatter(raw).unwrap_err();
        assert!(err.to_string().contains("Unterminated frontmatter"));
    }

    #[test]
    fn split_frontmatter_empty_file_errors() {
        let err = split_frontmatter("").unwrap_err();
        assert!(err.to_string().contains("empty"));
    }

    #[test]
    fn tool_list_splits_on_whitespace_handling_complex_tokens() {
        let mut m = manifest_for_test("x", "y", "user");
        m.allowed_tools = Some("Read Write Bash(git:*)  Edit".into());
        assert_eq!(
            m.tool_list(),
            vec!["Read", "Write", "Bash(git:*)", "Edit"]
                .into_iter()
                .map(String::from)
                .collect::<Vec<_>>(),
        );
    }

    #[test]
    fn tool_list_empty_when_field_missing() {
        let m = manifest_for_test("x", "y", "user");
        assert!(m.tool_list().is_empty());
    }
}
